
## Picture Phrase: NumPy Challenge


## Overview

A picture with a specific phrase written on it was taken and turned it into a NumPy array of RGB values. The array was then transformed using a set of specific NumPy transformations. It will be your task to transform the array back to the original by performing the same NumPy transformations. 

In the end, you will reshape the array and show the corresponding image associated with it..

The result of each question shown after each question cell.

### Important Details

- Please make sure you follow all the steps <u>*in order*</u> or the picture may not be viewable and the question may be counted incorrect in CodeGrade.  
- Since these operations must be done in order, you MUST use the variable from the previous question when performing the operation. For example:

   **SampleQ1:** Create a variable called SampleQ1 that is equal to 1.

       SampleQ1 = 1

   **SampleQ2:** Add 1.

      SampleQ2 = SampleQ1 + 1

   **SampleQ3:** Multiply by 2

      SampleQ3 = SampleQ2 * 2
      
- To prevent creating views of the array where you might inadvertently change an earlier array without knowing it, we have included code in most questions to create a copy of the previous array.  You will need to uncomment that code as you go along.  [See this link for more information on views vs copies.](https://www.w3schools.com/python/numpy/numpy_copy_vs_view.asp#:~:text=The%20main%20difference%20between%20a,will%20not%20affect%20the%20copy.)


## Standard Imports

In [1]:
# load standard imports
import numpy as np
from matplotlib import pyplot as plt
np.set_printoptions(suppress=True) #prevent numpy exponential 

## Load Array 

In [2]:
# load image
img = np.loadtxt('array.txt', dtype='U32')

In [3]:
# view array
img

array(['-7.45', '2484.4', '-6.7', ..., '42894.65', '40607.15', '46813.95'],
      dtype='<U32')

It doesn't look like it, but this array actually contains RGB values (albeit mixed up and transformed) for a picture. Let's check the shape of the data.

## NumPy Exercises

**Exercise:** What is the shape of the NumPy array?

In [ ]:
### ENTER CODE HERE ###

In [4]:
#answer

(812160,)

Next, let's check the data type.

**Exercise:** What is the data type for the array?

In [5]:
#answer

numpy.ndarray

### Questions

For the next tasks, save your output as the question number.  The first two have been done for you as examples.

In order to perform mathematical operations on the data, we will need the values to be in a numerical format. 

**Q1:** Update the data type of `img` to `float64` and call it `Q1`.

In [6]:
# update `img` to float and call it Q1
Q1 = img.astype('float64')
Q1

array([   -7.45,  2484.4 ,    -6.7 , ..., 42894.65, 40607.15, 46813.95])

One of the key features of NumPy is its ability to perform mathematical operations on whole blocks of data. Let's look at some of these operations.

**Q2:** Multiply each element in the `Q1` array by 10 and call this new array `Q2`.

In [7]:
# multiply each element by 10 
Q2 = Q1 * 10
Q2

array([   -74.5,  24844. ,    -67. , ..., 428946.5, 406071.5, 468139.5])

**Q3:** 
    
1. Create a NumPy array called `arr` of the following five values: 32, 54, 67, 88, 4
    
2. Use the NumPy's sum method to add all the values together and store as a variable called `total`.  
    
3. Add this total to every element of the `Q2` array and call this new array `Q3`.

In [8]:
#answer

array([   170.5,  25089. ,    178. , ..., 429191.5, 406316.5, 468384.5])

**Q4:** Add the `Q3` array to itself (example: array + array) and call it `Q4`. This will add the corresponding values in each cell to each other.

In [9]:
#answer

array([   341.,  50178.,    356., ..., 858383., 812633., 936769.])

**Q5:** Subtract 1 from the first item in the `Q4` array, 2 from the 2nd item, 3 from the 3rd item...n from the n<sup>th</sup> item and call the new array `Q5`. You should **not** do this using any type of loop as that wouldn't be efficient. 

**Hint:** Think about creating a second array that could be used in combination with the first array to complete the task.  NumPy's [linspace() method](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html) may come in handy here.

In [10]:
#answer

array([   340.,  50176.,    353., ...,  46225.,    474., 124609.])

**Code Check:**  To check your code so far, the output of your `Q5` array should be: 

`array([ 340., 50176., 353.,..., 46225., 474., 124609.])`.  

If your output does not match, go back and check your above steps.

**Q6:** 
- Uncomment the line below that copies the `Q5` array and calls the new array `Q6` (this step makes sure that we are working with a copy instead of a view of the previous array). 
- Update the 4th element in the `Q6` array to `46225`. <u>Careful</u>: Remember that Python is zero based index.

In [11]:
# create a copy
Q6 = Q5.copy()

In [12]:
#answer

array([   340.,  50176.,    353., ...,  46225.,    474., 124609.])

**Q7:** 
- Copy the `Q6` array and call the new array `Q7`. 
- Select the indices from `1,000` to `200,000` (including the element in the `200,000` index) and the indices from `400,000` to `600,000` (including the element in the `600,000` index) in `Q7` and add `255` to only those values.

In [13]:
# Create a copy
Q7 = Q6.copy()

In [14]:
#answer

array([   340.,  50176.,    353., ...,  46225.,    474., 124609.])

**Code Check:**  To check your code, confirm the following output from `Q7`:
```
Q7[999] should equal 46225.0
Q7[1000] should equal 349.0
Q7[600000] should equal 341.0
Q7[600001] should equal 50625.0
```

**Q8:** 
- Copy the `Q7` array and call the new array `Q8`.  
- Reshape the `Q8` array to a two-dimensional, two-column array.

In [15]:
# create a copy
Q8 = Q7.copy()

In [16]:
#answer

array([[   340.,  50176.],
       [   353.,  46225.],
       [   349.,  51984.],
       ...,
       [   340.,  50176.],
       [   353.,  46225.],
       [   474., 124609.]])

**Q9:** 
- Copy the `Q8` array and call the new array `Q9`.  
- Swap the two columns from this `Q9` array. In other words, the first column will now be the second and the second column will now be the first.

In [17]:
# create a copy
Q9 = Q8.copy()

In [18]:
#answer

array([[ 50176.,    340.],
       [ 46225.,    353.],
       [ 51984.,    349.],
       ...,
       [ 50176.,    340.],
       [ 46225.,    353.],
       [124609.,    474.]])

**Q10:** 
- Copy the `Q9` array and call the new array `Q10`. 
- Reverse the order of the rows of the `Q10` array.

In [19]:
# create a copy
Q10 = Q9.copy()

In [20]:
#answer

array([[124609.,    474.],
       [ 46225.,    353.],
       [ 50176.,    340.],
       ...,
       [ 51984.,    349.],
       [ 46225.,    353.],
       [ 50176.,    340.]])

**Code Check:** To check your code, the output of `Q10` should be the following:

```
array([[124609.,    474.],
       [ 46225.,    353.],
       [ 50176.,    340.],
       ...,
       [ 51984.,    349.],
       [ 46225.,    353.],
       [ 50176.,    340.]])
```

If your output does not match, go back and review your previous steps.

**Q11:** 
- Copy the `Q10` array and call it `Q11`. 
- Set `Q11`'s 101st row as `[46225, 353]`. <u>Careful</u>: Remember that Python is zero based index

In [21]:
# create a copy
Q11 = Q10.copy()

In [22]:
#answer

array([[124609.,    474.],
       [ 46225.,    353.],
       [ 50176.,    340.],
       ...,
       [ 51984.,    349.],
       [ 46225.,    353.],
       [ 50176.,    340.]])

**Q12:** 
- Copy the `Q11` array and call it `Q12`. 
- Set the **first element** in `Q12's` 100,001st row to `46656`.  The second element in that row should remain unchanged. <u>Careful</u>: Remember that Python is zero based index

In [23]:
# create a copy
Q12 = Q11.copy()

In [24]:
#answer

array([[124609.,    474.],
       [ 46225.,    353.],
       [ 50176.,    340.],
       ...,
       [ 51984.,    349.],
       [ 46225.,    353.],
       [ 50176.,    340.]])

**Q13:** 
- Copy the `Q12` array and call it `Q13`. 
- Subtract 125 from every element in the **second column** for `Q13`.

In [25]:
# create a copy
Q13 = Q12.copy()

In [26]:
#answer

array([[124609.,    349.],
       [ 46225.,    228.],
       [ 50176.,    215.],
       ...,
       [ 51984.,    224.],
       [ 46225.,    228.],
       [ 50176.,    215.]])

**Q14:** 
- Copy the `Q13` array and call it `Q14`. 
- Take the square root of every element in the **first column** of `Q14`.

In [27]:
# copy
Q14 = Q13.copy()

In [28]:
#answer

array([[353., 349.],
       [215., 228.],
       [224., 215.],
       ...,
       [228., 224.],
       [215., 228.],
       [224., 215.]])

**Code Check:** To check your code, the output of `Q14` should be:
```
array([[353., 349.],
       [215., 228.],
       [224., 215.],
       ...,
       [228., 224.],
       [215., 228.],
       [224., 215.]])
```     
If your output does not match, go back and review your previous steps.

**Q15:** 
- Copy the `Q14` array and call it `Q15`.  
- Change any value in the `Q15` array that is currently 999 and set it to 117.

In [29]:
# copy
Q15 = Q14.copy()

In [30]:
#answer

array([[353., 349.],
       [215., 228.],
       [224., 215.],
       ...,
       [228., 224.],
       [215., 228.],
       [224., 215.]])

# **Q16:** 
1. Copy the `Q15` array and call it `Q16`.
2. Take the [transpose](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.T.html#numpy.ndarray.T) of the `Q16` array.
3. Subtract 125 from the first column of all rows.
4. Take the transpose of the array again (this puts it back in its original state) making sure that all the changes are saved as `Q16`.

In [31]:
# copy
Q16 = 

In [32]:
#answer

array([[228., 224.],
       [215., 228.],
       [224., 215.],
       ...,
       [228., 224.],
       [215., 228.],
       [224., 215.]])

**Q17:** 
- Copy the `Q16` array and call it `Q17`.  
- Reshape the `Q17` array to (423, 640, 3). Note: the 3 values in each row are the RGB values for every pixel of the picture.

In [2]:
# copy
Q17 = 

In [34]:
#answer

array([[[228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.],
        ...,
        [228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.]],

       [[228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.],
        ...,
        [228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.]],

       [[228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.],
        ...,
        [228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.]],

       ...,

       [[ 10.,   9.,  14.],
        [ 10.,   9.,  14.],
        [ 10.,   9.,  14.],
        ...,
        [228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.]],

       [[ 13.,  12.,  17.],
        [ 12.,  11.,  16.],
        [ 11.,  10.,  15.],
        ...,
        [228., 224., 215.],
        [228., 224., 215.],
        [228., 224., 215.]],

       [[ 12.,  11.,  16.],
        [ 11.,  10.,  15.],
        [ 10.,  

**Q18:** 
- Copy the `Q17` array and call it `Q18`.  
- Cast the entire `Q18` array as an `int`.

In [35]:
# copy
Q18 = Q17.copy()

In [36]:
#answer

array([[[228, 224, 215],
        [228, 224, 215],
        [228, 224, 215],
        ...,
        [228, 224, 215],
        [228, 224, 215],
        [228, 224, 215]],

       [[228, 224, 215],
        [228, 224, 215],
        [228, 224, 215],
        ...,
        [228, 224, 215],
        [228, 224, 215],
        [228, 224, 215]],

       [[228, 224, 215],
        [228, 224, 215],
        [228, 224, 215],
        ...,
        [228, 224, 215],
        [228, 224, 215],
        [228, 224, 215]],

       ...,

       [[ 10,   9,  14],
        [ 10,   9,  14],
        [ 10,   9,  14],
        ...,
        [228, 224, 215],
        [228, 224, 215],
        [228, 224, 215]],

       [[ 13,  12,  17],
        [ 12,  11,  16],
        [ 11,  10,  15],
        ...,
        [228, 224, 215],
        [228, 224, 215],
        [228, 224, 215]],

       [[ 12,  11,  16],
        [ 11,  10,  15],
        [ 10,   9,  14],
        ...,
        [228, 224, 215],
        [228, 224, 215],
        [228, 224, 215]]

**Q19:** Uncomment the first line of code below (`img = Q18`).  Then run the code block to plot your array. 

In [4]:
# uncomment the line below
img = Q18

# # plot image
# fig = plt.figure(figsize=(15, 10))
# plt.imshow(img)
# plt.show()

**Q20:** Create a variable called `Q20` that is a string of the phrase in the picture (include punctuation but not the name of the author).  You can of course hard code this string.

In [38]:
#answer

When my cats aren't happy, I'm not happy. Not because I care about their mood but because I know they're just sitting there thinking up ways to get even.
